core/headerchain.go

```diff
@@ -307,6 +307,35 @@ func (hc *HeaderChain) GetBlockHashesFromHash(hash common.Hash, max uint64) []co
 	return chain
 }
 
+// GetAncestor retrieves the Nth ancestor of a given block. It assumes that either the given block or
+// a close ancestor of it is canonical. maxNonCanonical points to a downwards counter limiting the
+// number of blocks to be individually checked before we reach the canonical chain.
+//
+// Note: ancestor == 0 returns the same block, 1 returns its parent and so on.
+func (hc *HeaderChain) GetAncestor(hash common.Hash, number, ancestor uint64, maxNonCanonical *uint64) (common.Hash, uint64) {
+	if ancestor > number {
+		return common.Hash{}, 0
+	}
+	for ancestor != 0 {
+		if ancestor > 1 && rawdb.ReadCanonicalHash(hc.chainDb, number) == hash {
+			number -= ancestor
+			return rawdb.ReadCanonicalHash(hc.chainDb, number), number
+		}
+		if *maxNonCanonical == 0 {
+			return common.Hash{}, 0
```

Comment:
This will trigger for `skip == 0` and number of requests > 192.

Reply:
All in all I think it's too convoluted that you have special case handling for `skip == 0` merged in with the other case seamlessly. If `skip == 0` is to be handled explicitly, please do so in a prior loop. I think it's much cleaner to have the two cases separate than together.

Reply:
I used the same downloader.MaxHeaderFetch that limits the number of headers retrieved.

Reply:
That's a client side limit, not a server side one. There was no such limit beforehand.

Reply:
You might be right about the skip==0 case though. I'm changing it now.

Reply:
Also why convolute the concepts of max header count with max canonical headers? They have nothing to do with each other.

Reply:
The only reason is the case you just mentioned. Ugly solution, I know :) I'll push the new version in a few minutes.

Reply:
done